server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1085,6 +1120,7 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
      */
     public CanMatchResponse canMatch(ShardSearchRequest request) throws IOException {
         assert request.searchType() == SearchType.QUERY_THEN_FETCH : "unexpected search type: " + request.searchType();
+        assert request.readerId() == null : "request with reader_id bypass can_match phase";
```

Comment:
Is it temporary ? We've said that it should be possible to run the can_match phase as long as we retrieve the reader and extend the keep alive if the request cannot match ?

Reply:
Yes. Once your PR is in, I will update this.

Reply:
ok, thanks for confirming

Reply:
I added a TODO: https://github.com/elastic/elasticsearch/pull/55265/commits/ab7086bd534c673e4b5e7945bee43cc87c0b71f4